### PR TITLE
CMakeLists.txt and test cleanup

### PR DESCRIPTION
**Pull request by eidheim**
* Added source_clang_test to figure out how the build should look like
* In order to avoid duplicate compilation of files, project_shared_sources library was added that is used in both ./src and ./tests. This drastically improves the compilation speed of the tests.
* General cleanup of CMakeLists.txt's to avoid duplicate code in src/CMakeLists.txt and tests/CMakeLists.txt

### CODE REVIEW

**Comment by eidheim** [comment]
@zalox could you have a look at this? We should merge this soon so I can continue to make tests. 

By the way, CI is awesome!:)

**Comment by eidheim** [comment]
@zalox Do you know why Travis CI was not run on the last commit above?

**Comment by zalox** [comment]
I see no reason for why the CI didn't build. The only thing I've noticed is that it won't run if there are syntax errors in `.travis.yml`.

I'm reviewing now :)

File: tests/source_clang_test.cc [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+//broadwayd&
+//make test
+
+int main() {
+  auto app=Gtk::Application::create();
+  Gsv::init();
+  
+  Config::get().project.default_build_path="./build";
+  Source::ClangView *clang_view=new Source::ClangView(boost::filesystem::canonical(std::string(JUCI_TESTS_PATH)+"/clang_project/main.cpp"),
+                                                      Gsv::LanguageManager::get_default()->get_language("cpp"));
```
Comment (zalox) [inline tests/source_clang_test.cc]: Raw pointer for fast destruction?

Comment (eidheim) [inline tests/source_clang_test.cc]: Yeah the destruction for Source::ClangView is a bit special due to libclang.

File: src/source_clang.cc [hunk 2]
```diff
@@ -1112,18 +1112,16 @@ Source::ClangViewAutocomplete(file_path, language) {
...
-    auto iter=get_iter_for_dialog();
-    selection_dialog=std::unique_ptr<SelectionDialog>(new SelectionDialog(*this, get_buffer()->create_mark(iter), true, true));
-    auto rows=std::make_shared<std::unordered_map<std::string, clang::Offset> >();
-    auto methods=clang_tokens->get_cxx_methods();
-    if(methods.size()==0)
-      return;
-    for(auto &method: methods) {
+    auto cxx_methods=clang_tokens->get_cxx_methods();
+    if(cxx_methods.size()==0)
+      return methods;
```
Comment (zalox) [inline src/source_clang.cc]: I think this check is redundant, the for-loop below does essentially the same thing. I realise this wasn't added by this commit though.

Comment (zalox) [inline src/source_clang.cc]: I just realised it became redundant in this commit.

File: tests/source_clang_test.cc (hunk 1 above)
Comment (eidheim) [inline tests/source_clang_test.cc]: You are right, I'll fix it later.

File: src/source_clang.cc (hunk 2 above)
Comment (eidheim) [inline src/source_clang.cc]: You are right, the if statement should be removed.

Comment (eidheim) [inline src/source_clang.cc]: Fixed!

**Pull request merged**